{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lassa data set evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def histograms(stats):\n",
    "    # Creating histograms with custom labels\n",
    "    fig, axes = plt.subplots(1, 4, figsize=(10, 3))  # Adjust grid if you have a different layout\n",
    "\n",
    "    stats[['N_share', 'Length', 'RelativeLength', 'RelativeError']].hist(\n",
    "        bins=50, log=True, ax=axes\n",
    "    )\n",
    "\n",
    "    # Setting y-axis label for the left column\n",
    "    axes[0].set_ylabel('Number of sequences')\n",
    "\n",
    "    axes[0].set_xlim(0, 1)\n",
    "    axes[1].set_xlim(left=0)\n",
    "    axes[2].set_xlim(left=0)\n",
    "    axes[3].set_xlim(0, 1)\n",
    "\n",
    "    # Setting x-axis labels with the feature names\n",
    "    features = ['N_share', 'Length', 'RelativeLength', 'RelativeError']\n",
    "    for ax, feature in zip(axes.flatten(), features):\n",
    "        ax.set_xlabel(feature)\n",
    "        ax.set_title('')  # Remove the title above each plot\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "fname_stats = ''\n",
    "fname_sequences = ''\n",
    "outdir = ''\n",
    "\n",
    "minlen = 2000\n",
    "max_n_share = 0.01\n",
    "min_relative_length = 0.95\n",
    "min_ref_cov = 0.95"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data ingestion.\n",
    "\n",
    "We compute the following features\n",
    "- the relative length of the sequence compared to the reference\n",
    "- edit distance in relation to the match length\n",
    "- the amount of the reference covered by the alignment\n",
    "- the amount of the query covered by the alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats = pd.read_csv(fname_stats, sep='\\t')\n",
    "seqstats['RelativeLength'] = seqstats['Length'] / seqstats['ReferenceLength']\n",
    "seqstats['RelativeError'] = seqstats['EditDistance'] / (seqstats['ReferenceEnd'] - seqstats['ReferenceStart'])\n",
    "seqstats['ReferenceCoverage'] = (seqstats['ReferenceEnd'] - seqstats['ReferenceStart']) / seqstats['ReferenceLength']\n",
    "seqstats['QueryCoverage'] = (seqstats['QueryEnd'] - seqstats['QueryStart']) / seqstats['Length']\n",
    "\n",
    "seqstats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overall length and N_share distribution before any filtering has happened."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats[['Length', 'N_share']].hist(bins=50, log=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing unaligned sequences.\n",
    "\n",
    "We remove all sequences that were not mapped. Let's have a look at the 10 largest sequences first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped = seqstats[seqstats['IsForward'].isna()][['Sequence', 'Length', 'N_share']]\n",
    "print(f'Number of unmapped reads={len(unmapped)}')\n",
    "unmapped.nlargest(10, ['Length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped.hist(bins=20, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_mapped = seqstats.dropna(subset='Reference')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing supplementary alignments\n",
    "\n",
    "Additional to the best hit, minimap2 also sometimes computes supplementary alignments. We remove them here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_primary = seqstats_mapped[seqstats_mapped['IsSupplementaryAlignment'] == False]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove short sequences and those with too many Ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_minlen = seqstats_mapped[seqstats_mapped['Length'] >= minlen]\n",
    "seqstats_high_coverage = seqstats_minlen[seqstats_minlen['N_share'] <= max_n_share]\n",
    "\n",
    "print('Before filtering:', len(seqstats_mapped))\n",
    "print(f'N-share larger than {max_n_share}:', len(seqstats_mapped[seqstats_mapped['N_share'] > max_n_share]))\n",
    "print(f'Length smaller than {minlen}:', len(seqstats_mapped[seqstats_mapped['Length'] < minlen]))\n",
    "print('After filtering:', len(seqstats_high_coverage))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove too short genomes\n",
    "We remove the genomes that are too short compared to the reference genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_long_seqs = seqstats_high_coverage[seqstats_high_coverage['RelativeLength'] > min_relative_length]\n",
    "print(\n",
    "    f'{len(seqstats_long_seqs)} of {len(seqstats_high_coverage)} sequences '\n",
    "    f'remain after filtering for {min_relative_length} minimum relative length'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally we remove all sequences, where the alignment does not cover enough of the target genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_long_seqs.nsmallest(10, ['ReferenceCoverage'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqstats_min_ref_cov = seqstats_long_seqs[seqstats_long_seqs['ReferenceCoverage'] >= min_ref_cov]\n",
    "print(f'Keeping {len(seqstats_min_ref_cov)} of {len(seqstats_long_seqs)} sequences')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per reference genome counts\n",
    "Let's have a look what is left after filtering with respect to the distinct reference genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_counts_raw = seqstats_primary['Reference'].value_counts()\n",
    "ref_counts_minlen = seqstats_minlen['Reference'].value_counts()\n",
    "ref_counts_high_cov = seqstats_high_coverage['Reference'].value_counts()\n",
    "ref_counts_filtered = seqstats_min_ref_cov['Reference'].value_counts()\n",
    "pd.DataFrame({'Raw': ref_counts_raw, 'MinLen': ref_counts_minlen, 'MaxN': ref_counts_high_cov,  'Filtered': ref_counts_filtered})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "histograms(seqstats_min_ref_cov)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read orientation\n",
    "Finally, lets get the filtered reads with harmonized orientation sorted into the Segments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "\n",
    "df_filtered_final = seqstats_min_ref_cov\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "segments = sorted(set(df_filtered_final['Segment'])) \n",
    "fname_template_out = os.path.join(outdir, \"{}.fasta\")\n",
    "\n",
    "outfiles = {\n",
    "    segment: open(fname_template_out.format(segment), 'w')\n",
    "    for segment in segments\n",
    "}\n",
    "\n",
    "segment_and_orientation = {\n",
    "    row['Sequence']: (row['IsForward'], row['Segment'])\n",
    "    for _, row in df_filtered_final.iterrows()\n",
    "}\n",
    "\n",
    "for record in SeqIO.parse(fname_sequences, \"fasta\"):\n",
    "    if record.id not in segment_and_orientation:\n",
    "        continue\n",
    "    is_forward, segment = segment_and_orientation[record.id]\n",
    "    if is_forward is None:\n",
    "        raise RuntimeError(f\"Invalid orientation {is_forward}\")\n",
    "    orientation = 'forward' if is_forward else 'reverse'\n",
    "    new_record = SeqRecord(\n",
    "        seq=record.seq if is_forward else record.seq.reverse_complement(),\n",
    "        id=record.id,\n",
    "        name=record.name,\n",
    "        description=f'{record.description}|{orientation}|{segment}'\n",
    "    )\n",
    "    SeqIO.write(new_record, outfiles[segment], \"fasta\")\n",
    "\n",
    "for outfile in outfiles.values():\n",
    "    outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hits_merge = seqstats_mapped.groupby('Sequence').agg(\n",
    "#     N_alignments=('IsForward', 'count'),\n",
    "#     SameOrientation=('IsForward', lambda x: x.nunique() == 1),\n",
    "#     SameReference=('Reference', lambda x: x.nunique() == 1),\n",
    "#     SequenceLength=('Length', 'max'),\n",
    "# ).reset_index()\n",
    "# #hits_merge[(hits_merge['SameOrientation'] == False) | (hits_merge['SameReference'] == False)]\n",
    "# hits_merge[hits_merge['N_alignments'] > 1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
